### ISSUE TITLE
Audit: MANETH-41

### ISSUE THREAD

**Issue by 0xKitsune** [manifoldfinance/mevETH2#118]
## Status
Reported

## Type
Vulnerability

## Severity
Highest

## Code Snippet:
```js
function grantRewards() external payable {
    // Process the withdrawal queue, paying out any pending withdrawal tickets before updating the fraction.
    processWithdrawalQueue();
    if (!(msg.sender == address(stakingModule) || msg.sender == mevEthShareVault)) revert MevEthErrors.InvalidSender();
 
    /// @dev Note that while a small possiblity, it is possible for the MevEthShareVault rewards + fraction.elastic to overflow a uint128.
    /// @dev in this case, the grantRewards call will fail and the funds will be secured to the MevEthShareVault.beneficiary address.
    fraction.elastic += uint128(msg.value);
    emit Rewards(msg.sender, msg.value);
}
```

## Remediation
	
The root of the issue is how rewards distribution is currently designed and also a common challenge with liquid staking protocols.

The share rate should be based on the existing funds as well as any unrealised profit from validator nodes. This is similar to how lending protocols should take any outstanding debt and accumulated interest into account even if the loan has not yet been repaid.

The unrealised profit could be calculated from an APR based on all previous rewards or on a constant rate multiplied by the number of validators (and assuming they do not get penalised on the consensus layer).

## Description 
```

MevEth.sol:grantRewards (L323-332)

The staking protocol obtains rewards from MEV and node operators validating blocks on the consensus layer. Rewards are accumulated in the MevEthShareVault contract. Then, the operator of the contract calls MevEthShareVault.sol:payRewards to submit a batch of rewards to MevEth via grantRewards.

This process batches rewards together in one single update and therefore does not linearly divide rewards among stakers. The MevEth contracts also allows for instant deposits and withdrawals (flash-staking).

It becomes possible for a user to steal the rewards (depending on their capital) without taking on any risk:

1. Front-run or sandwich the call to payRewards
2 .Call MevEth:deposit with a very large amount of ETH.
3. After execution of payrewards
4. Call MevEth:withdraw to re-obtain the initial stake plus a large chunk of the rewards depending on the initial stake amount.

In other words, the user briefly becomes the user with the largest stake and will receive a portion of the rewards equal to that stake percentage. They can then immediately unstake again.

The user is risk-free from any risk that a normal taker would be exposed to, such as validator slashing and protocol exploits.
```
